### PR TITLE
SAM BOSSA bootloader support and CDC-ACM line rate change detection

### PR DESCRIPTION
**Pull request by nzmichaelh**
The BOSSA bootloader is used on Arduino compatible SAM boards like the Arduino MKR Zero, Adafruit Trinket M0, and Seeedstudio XIAO.  These boards use the 'Arduino hack' of setting the USB CDC baud rate to 1200 baud to signal the application to restart into the bootloader.

Add a new callback to the CDC-ACM UART API so the application can listen for baud rate changes.  Add a SAM0 specific module that listens for these changes and restarts into the bootloader if found.

With these changes, a "west flash" will reset the board and flash the new version.

### CODE REVIEW

File: subsys/usb/class/cdc_acm.c
Comment (jfischer-phytec-iot) [inline subsys/usb/class/cdc_acm.c:939]: -EINVAL or -EFAULT?

File: boards/arm/adafruit_trinket_m0/doc/index.rst
Comment (jfischer-phytec-iot) [inline boards/arm/adafruit_trinket_m0/doc/index.rst:102]: This somehow implies that USB device stack and CDC ACM is always present (even if one would read the text below).

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:939]: EINVAL SGTM, done.

File: boards/arm/adafruit_trinket_m0/doc/index.rst
Comment (nzmichaelh) [reply boards/arm/adafruit_trinket_m0/doc/index.rst:102]: Reworded.  PTAL.

File: subsys/usb/class/cdc_acm.c
Comment (pabigot) [inline subsys/usb/class/cdc_acm.c:943]: This looks a little odd to me: you can pass a null callback, but only if no callback has been set.  Once a callback is set, you can't clear it.

If the intent that this function can be invoked only once, then that should be documented.  If it's possible to set the callback multiple times then a lock must be provided to ensure that use of the callback is consistent with a configured pair of callback and userdata.

Without knowing much about this I think the simplest solution is document that the function can be successfully invoked only once, and that passing a null callback returns `-EINVAL` while passing a non-null callback if one has already been set returns `-EBUSY` or `-EALREADY`.  `user_data` can be null or not, but you can't change it once the callback has been set.

If that doesn't satisfy the use case document and implement something that does.

**Review by pabigot — DISMISSED** [review]
I can't speak to the BOSSA stuff, but the API extension needs some clarification.

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:943]: Hi Peter.  I had a look at the other callback implementations and you're right - any callback set will overwrite any previous callback set.

Changed this to match.

**Comment by carlescufi** [comment]
@pabigot please revisit.

**Review by pabigot — COMMENTED** [review]
The new callback still isn't managed in a thread-safe way, but neither is the existing write callback.  I don't think fixing all that is in scope for this.  So I'll remove my block, but I won't approve it at this time.  Assuming a UART/USB maintainer thinks this is OK I'll willing to second it.

**Comment by nzmichaelh** [comment]
Hi @jfischer-phytec-iot I see you changed the title to be UART API specific.  Should I split out the BOSSA changes then?

**Comment by nzmichaelh** [comment]
Hi @jfischer-phytec-iot @finikorg I see you're the USB subsystem owners.  What do you think of the API change?  If required, I'm happy to do a follow up PR that changes the CDC ACM implementation to register callbacks in a thread safe way.  See @pabigot 's comment for more.

**Comment by jfischer-phytec-iot** [comment]
> Hi @jfischer-phytec-iot I see you changed the title to be UART API specific. Should I split out the BOSSA changes then?

I think not or not yet, my intention was to get more attention.

>What do you think of the API change?

+1, but I think we should discuss it in the API meeting. @carlescufi

File: include/drivers/uart.h
Comment (anangl) [inline include/drivers/uart.h:1166]: Since this API function is optional, it would be good to return `-ENOTSUP` when a given driver does not implement it.

**Review by pabigot — COMMENTED** [review]
This seems a little under-defined.  AFAICT all you get as a user is a callback suggesting that something changed, and the user has to guess what caused the callback.

At this time the only use case is for a baud rate change.  Is there a reason to invoke it on DCD change, or CTS, or RI, or other signals? If so, does it make sense to pass to the callback the bits identifying what changed?  Then the API should document what line control state change detections are required when this feature is implemented in a driver, and which are optional.

If not, should this just be a baud rate changed callback, and pass the new baud rate as a parameter?

**Comment by pabigot** [comment]
Notes from 2020-06-30 API telecon.  Edit if I missed something.

JF: Arduino BOSSA uses changes of baud rate to determine whether system should trigger boot loader.  Do we need this?

KC: Seems limited

JF: Likely that other signals than baud rate could be used.

PB: Not sure hardware would support notification of change in baud rate even if autobaud supported.  Similar for low-level control signals.

Summary: Need more input to determine whether this is useful, what would be useful, and whether any of it is implementable on anything other than CDC-ACM.  If not this may be suitable as a device-specific extension for CDC-ACM somehow.

**Comment by nzmichaelh** [comment]
Hi Peter.  There's two PRs for adding BOSSA / Arduino reset support so far - this and #25422

#25422 directly modifies cdc_acm and the feedback was that it was too tightly coupled and should be a generic API.  This PR is a generic API, but the feedback is that there's not enough use cases to add it.

We're getting conflicting feedback.  How should we proceed?

**Comment by jfischer-phytec-iot** [comment]
> If not this may be suitable as a device-specific extension for CDC-ACM somehow.

> We're getting conflicting feedback. How should we proceed?

I guess something was misunderstood. From my side, only this proposal makes sense, not the hack in #25422

**Comment by pabigot** [comment]
> We're getting conflicting feedback. How should we proceed?

Sorry; like many things in Zephyr what at first seems simple gets complicated.

My position is that if we're adding new API to the generic UART driver we need to know the requirements for the API.  `uart_line_ctrl_callback_set()` needs justification, more clear definition, and confirmation that it can be implemented on multiple platforms.

I understand that the goal is:

> Detect an Arduino-style upgrade request and implement the required response

The Zephyr implementation should respect these distinct phases.  I agree with @jfischer-phytec-iot that the approach in #25422 is not acceptable: it only works with one platform on one driver, and puts target-specific upgrade logic in a serial driver.  If somebody has a non-SAM0 Arduino-style board it won't help them, and maintaining the serial driver becomes complex.

So my first question is: Can any UART driver other than CDC-ACM detect this event?  If so, what other line-control events can be detected, and how must the subsystem `uart_something_changed()` API be defined to support them?

In answering that I've now learned about [BOSSA](http://www.shumatech.com/web/products/bossa) which is a red-herring, and  about how original Arduino hardware [tied DTR to the MCU RESETn signal](https://liudr.wordpress.com/2016/02/26/how-arduino-bootloader-works/) to force a reboot, and then how for USB where DTR is not a physical signal this is implemented by [opening the serial port at 1200 baud and then closing it](https://arduino.stackexchange.com/a/32737).

So it appears that this functionality is only ever useful for CDC_ACM UARTs, and for wired RS-232 there's no software needed because the reset is done in hardware.

So: IMO the line state change detection API doesn't belong in the UART API.

In that case, I would propose to add an an option `CONFIG_CDC_ACM_ENABLE_ARDUINO_RESET` flag, which puts the detection logic into the cdc_acm driver and defines a non-syscall `cdc_acm_arduino_reset_callback(dev, cb)` function that installs a callback to be used when the event occurs.  You'll need to add a small header in `install/drivers/uart/cdc_acm.h` to provide the definition.

Reading the links above I'm not convinced the response to the event needs to erase the first word of the application: it seems it's just supposed to be a soft reset, which starts the bootloader, which may or may not perform an update.  I may be wrong in that interpretation.

But in any case the most flexible approach is just detecting the situation with API like the one two paragraphs up, and letting the application determine what "reset" means.  (If necessary I suppose a `lib/arduino` could be added with `arduino_reset_to_upgrade()` implemented for shared use, but that should be a separate PR from adding the ability to detect the request.)

@jfischer-phytec-iot Does this plan make sense to you?

**Comment by nzmichaelh** [comment]
> Sorry; like many things in Zephyr what at first seems simple gets complicated.

It's all good.  Thanks for the detailed reply.

> My position is that if we're adding new API to the generic UART driver we need to know the requirements for the API. `uart_line_ctrl_callback_set()` needs justification, more clear definition, and confirmation that it can be implemented on multiple platforms.
> 
> I understand that the goal is:
> 
> > Detect an Arduino-style upgrade request and implement the required response
> 
> The Zephyr implementation should respect these distinct phases. I agree with @jfischer-phytec-iot that the approach in #25422 is not acceptable: it only works with one platform on one driver, and puts target-specific upgrade logic in a serial driver. If somebody has a non-SAM0 Arduino-style board it won't help them, and maintaining the serial driver becomes complex.
> 
> So my first question is: Can any UART driver other than CDC-ACM detect this event? If so, what other line-control events can be detected, and how must the subsystem `uart_something_changed()` API be defined to support them?

Unfortunately I don't have other concrete use cases.  Other serial emulation protocols such as Bluetooth SPP or RFC2217 have similar features where the line control and configuration is set remotely.

I can imagine using this for a Zephyr based USB -> Serial adapter, but that's a hypothetical.
 
> In answering that I've now learned about [BOSSA](http://www.shumatech.com/web/products/bossa) which is a red-herring, and about how original Arduino hardware [tied DTR to the MCU RESETn signal](https://liudr.wordpress.com/2016/02/26/how-arduino-bootloader-works/) to force a reboot, and then how for USB where DTR is not a physical signal this is implemented by [opening the serial port at 1200 baud and then closing it](https://arduino.stackexchange.com/a/32737).
> 
> So it appears that this functionality is only ever useful for CDC_ACM UARTs, and for wired RS-232 there's no software needed because the reset is done in hardware.

Correct.  We can ignore the RS-232 case as well as no modern boards use it.
 
> So: IMO the line state change detection API doesn't belong in the UART API.
> 
> In that case, I would propose to add an an option `CONFIG_CDC_ACM_ENABLE_ARDUINO_RESET` flag, which puts the detection logic into the cdc_acm driver and defines a non-syscall `cdc_acm_arduino_reset_callback(dev, cb)` function that installs a callback to be used when the event occurs. You'll need to add a small header in `include/drivers/uart/cdc_acm.h` to provide the ~definition~ declaration.

SGTM.

BTW, another alternative is adding a `SYS_REBOOT_INTO_BOOTLOADER` in `power/reboot.h`, but then cdc_acm has specific knowledge about the Arduino bootloader.
 
> Reading the links above I'm not convinced the response to the event needs to erase the first word of the application: it seems it's just supposed to be a soft reset, which starts the bootloader, which may or may not perform an update. I may be wrong in that interpretation.

I agree.  Recent bootloaders will stay in the bootloader if there's a magic value at the top of RAM, and this PR uses that method.
 
> But in any case the most flexible approach is just detecting the situation with API like the one two paragraphs up, and letting the application determine what "reset" means. (If necessary I suppose a `lib/arduino` could be added with `arduino_reset_to_upgrade()` implemented for shared use, but that should be a separate PR from adding the ability to detect the request.)

The reset code is very SOC specific and also tiny, so I'd duplicate it for other non-SAM0 SOCs if it's ever needed.
 
> @jfischer-phytec-iot Does this plan make sense to you?

**Comment by nzmichaelh** [comment]
Changed to a CDC ACM specific callback.  PTAL.

**Review by jfischer-no — DISMISSED** [review]

File: include/usb/class/usb_cdc.h
Comment (pabigot) [reply include/usb/class/usb_cdc.h:175]: But "arduino reset" is an event that, based on [my analysis](https://github.com/zephyrproject-rtos/zephyr/pull/26173#issuecomment-652353731)  can only be detected in the CDC ACM driver.

> It is similar attempt as #25422 and I will not agree with any attempt to add a callback in the control request processing.

Do you have an alternative method to detect the event of a baud rate change to 1200?

File: subsys/usb/class/cdc_acm.c
Comment (pabigot) [reply subsys/usb/class/cdc_acm.c:242]: I should be clear: I'm fine with generalizing this from detecting a change to 1200 baud to detecting a change in baud rate from the last observed (or configured) value.  But unless there's evidence this can be done on other UART platforms I believe this should be a CDC-ACM--specific extension, and I don't see any reason to generalize from baud rate change to arbitrary line control change.

File: include/usb/class/usb_cdc.h
Comment (jfischer-phytec-iot) [reply include/usb/class/usb_cdc.h:175]: Not relevant, this is a header for CDC (ACM) class specification, whatever arduino-callback is not part of it.

Comment (pabigot) [reply include/usb/class/usb_cdc.h:175]: There I agree with @jfischer-phytec-iot.  I asked that this be put in `include/drivers/uart/cdc_acm.h` which is the correct path for driver-specific extensions to the CDC-ACM UART driver.

If this moves, are you more willing to accept this driver-specific extension?

File: soc/arm/atmel_sam0/common/bossa.c
Comment (pabigot) [inline soc/arm/atmel_sam0/common/bossa.c:44]: How do you know this is the right device to use?  Possibly replace by `CONFIG_UART_CONSOLE_ON_DEV_NAME` as there doesn't appear to be any coupling between devicetree identifiers and USB devices.

Then check whether the reset callback succeeds.

File: include/usb/class/usb_cdc.h
Comment (pabigot) [inline include/usb/class/usb_cdc.h:187]: This needs an `int` return value that can be provide `-ENOSYS`  (or `-ENOTSUP`) if something goes wrong.

File: subsys/usb/class/cdc_acm.c
Comment (pabigot) [inline subsys/usb/class/cdc_acm.c:762]: Verify that `dev->driver_api` is equal to `cdc_acm_driver_api` and return an error if it isn't.  Otherwise passing the wrong device will corrupt something else's state.

**Review by pabigot — CHANGES_REQUESTED** [review]
I'm OK with this if the CDC-ACM API extensions are moved to `include/drivers/uart/cdc_acm.h` and the driver passed to the extension is checked for validity.

File: include/usb/class/usb_cdc.h
Comment (nzmichaelh) [reply include/usb/class/usb_cdc.h:187]: Done.

Note that I modeled this on the other callback setters which also return void.

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:242]: As an alternative, I've changed this to `cdc_acm_dte_rate_callback_set` that is called when the USB host changes the baud rate.
The only use case is resetting into the bootloader, but now the CDC ACM code has nothing Arduino specific in it.

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:762]: Done.

File: include/drivers/uart.h
Comment (nzmichaelh) [reply include/drivers/uart.h]: Superseded

File: include/usb/class/usb_cdc.h
Comment (nzmichaelh) [reply include/usb/class/usb_cdc.h:175]: Moved

File: soc/arm/atmel_sam0/common/bossa.c
Comment (nzmichaelh) [reply soc/arm/atmel_sam0/common/bossa.c:49]: The console alias is no good, as it's common to have the console on a physical UART.

Leaving open while I think about it.

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c]: PTAL

**Review by pabigot — CHANGES_REQUESTED** [review]
OK; I like this better as it's more general.  One fix for byte order requested.

**Comment by tbursztyka** [comment]
@nzmichaelh Can you edit your PR title?

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:236]: Good catch.  I haven't tested this yet.  My plan is to wait on feedback on the structure before going further.

Leaving open until tested.

**Comment by pabigot** [comment]
> @nzmichaelh Can you edit your PR title?

Good catch; the previous text was confusing given the current content.  I also updated the summary slightly to clarify that this is no longer proposing a change to the UART subsystem API, just extending CDC-ACM UART device.

File: subsys/usb/class/cdc_acm.c
Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

@jfischer-phytec-iot what is the argument here against this current approach?

Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: >It is pure generic code, and then there is this very specific to a platform(IDE?) and architecture. IMHO, we should generally not allow such a thing. Which specific IDE bootloader method is next?

No, this is actually the only reliable way a USB CDC ACM implementation can detect that a Host has opened a serial port, as far as I know.

Comment (pabigot) [reply subsys/usb/class/cdc_acm.c]: Note that the use case and implementation here does not detect that a host has opened a serial port; it detects that the host has opened a serial port *at a different baud rate*.  I suppose this could be converted to a "serial-port-opened" callback, if there's a reason to provide that.

I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

Comment (jfischer-phytec-iot) [reply subsys/usb/class/cdc_acm.c]: > I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

You can use DTR for that, see samples/subsys/usb/cdc_acm

File: include/drivers/uart/cdc_acm.h
Comment (jfischer-phytec-iot) [inline include/drivers/uart/cdc_acm.h:29]: s/DTE rate change event/dwDTERate SetLineCoding request
s/virtual baud/baud/g

File: include/drivers/uart/cdc_acm.h
Comment (jfischer-phytec-iot) [inline include/drivers/uart/cdc_acm.h:26]: s/cdc_acm_dte_rate_callback_t/cdc_dte_rate_callback_t/g

File: subsys/usb/class/cdc_acm.c
Comment (jfischer-phytec-iot) [reply subsys/usb/class/cdc_acm.c]: > I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

> 
> @jfischer-phytec-iot what is the argument here against this current approach?

Generally I find this approach to go into bootloader mode badly designed. It is not that there are no other methods, for example USB DFU. Here whatever code is executed in the context of USB controll request, for me it is not right, but I probably would have no problem to use simple IPC (like DBUS) for that. This code is inserted and needs to be maintained, just for the reason that some IDE can trigger the bootloader for some types of boards without pressing the button. By the way, it only works if USB is included in the firmware.
That is why I do not like this approach. I have left some comments, otherwise I leave the decision if it should go in or not to others.

Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c]: Thanks Johann.  To clarify - this isn't for a new bootloader design, but for compatibility with the manufacturer's default bootloader.  I agree if we were starting from a greenfield then something like DFU is better.

File: subsys/usb/class/cdc_acm.c
Comment (nzmichaelh) [reply subsys/usb/class/cdc_acm.c:236]: Done and tested on the Trinket M0

File: include/drivers/uart/cdc_acm.h
Comment (nzmichaelh) [reply include/drivers/uart/cdc_acm.h:26]: Done

File: include/drivers/uart/cdc_acm.h
Comment (nzmichaelh) [reply include/drivers/uart/cdc_acm.h]: Done

**Comment by nzmichaelh** [comment]
Thanks for all the comments everyone.  The remaining issue is identifying the CDC ACM device name, but otherwise any objections?

File: soc/arm/atmel_sam0/common/bossa.c
Comment (pabigot) [reply soc/arm/atmel_sam0/common/bossa.c:49]: How about a Kconfig `BOSSA_UART_LABEL`, defaulted to `"CDC_ACM_0"`?   (That the label is always for a CDC-ACM UART doesn't change the fact that from the perspective of the protocol it's a UART, though the limitation could be noted in the help section of the option.)

File: subsys/usb/class/cdc_acm.c
Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: >You can use DTR for that, see samples/subsys/usb/cdc_acm

I am not sure that works on all platforms @jfischer-phytec-iot. From experience with J-Link's CDC ACM, I have found that only a baud rate **change** is reliable in getting through on all OSs. CC @emob-nordic. Maybe that's why this BOSSA bootloader uses this approach.

Unfortunately we don't have a system-wide notification message today so that users can subscribe and listen to baud rate (or any other hardware parameters) notifications. In the meantime, I would argue that this has the potential to be useful for others and I'd vote for its inclusion.

Comment (jfischer-phytec-iot) [reply subsys/usb/class/cdc_acm.c]: DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

**Comment by carlescufi** [comment]
> Thanks for all the comments everyone. The remaining issue is identifying the CDC ACM device name, but otherwise any objections?

That's correct, thanks!

File: subsys/usb/class/cdc_acm.c
Comment (carlescufi) [reply subsys/usb/class/cdc_acm.c]: > DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

I agree it's a misuse, but it's a practical misuse, a quirk.

Comment (jfischer-phytec-iot) [reply subsys/usb/class/cdc_acm.c]: Yeah, like running a red light.

File: soc/arm/atmel_sam0/common/bossa.c
Comment (nzmichaelh) [reply soc/arm/atmel_sam0/common/bossa.c:49]: I went for `_DEVICE_NAME` to match the style of `CONFIG_USB_CDC_DEVICE_NAME`

File: Kconfig.zephyr
Comment (pabigot) [inline Kconfig.zephyr:412]: This should be `BOOTLOADER_BOSSA_DEVICE_NAME` (or `..._DEV_NAME`, which seems a little more common.

```
git grep -B1 -i 'string.*device name'
```

Fixup will also need to be squashed into the introducing commit.

**Review by pabigot — CHANGES_REQUESTED** [review]

File: Kconfig.zephyr
Comment (nzmichaelh) [reply Kconfig.zephyr:412]: Done.  Kept DEVICE_NAME to match CDC_ACM_DEVICE_NAME

**Review by pabigot — APPROVED** [review]

File: soc/arm/atmel_sam0/common/bossa.c
Comment (pabigot) [inline soc/arm/atmel_sam0/common/bossa.c:60]: FYI this not being `CONFIG_BOOTLOADER_BOSSA_DEVICE_NAME` is why compliance check failed (this is an unrecognized Kconfig name).

Comment (nzmichaelh) [reply soc/arm/atmel_sam0/common/bossa.c:60]: Done

**Comment by nzmichaelh** [comment]
Hi folks.  The CDC ACM device name issue has been addressed and [1] defers the decision on this landing to others.

What's the next step here?

[1]: https://github.com/zephyrproject-rtos/zephyr/pull/26173#discussion_r452374618

File: Kconfig.zephyr
Comment (nashif) [inline Kconfig.zephyr:402]: I feel we do need some other place for all those bootloader options, the top level Kconfig.zephyr file does not seem to be the right place for this.

You are just adding into something that already exists, so we can fix this later I guess.

**Review by nashif — APPROVED** [review]

**Comment by carlescufi** [comment]
@jfischer-phytec-iot can you please approve?

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
